kame/kame/dns6conv/dns6conv.1
.\"	$KAME: dns6conv.1,v 1.4 2002/03/27 03:01:35 jinmei Exp $
.\"
.\" Copyright (C) 2001 WIDE Project.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd January 13, 2001
.Dt DNS6CONV 1
.Os KAME
.\"
.Sh NAME
.Nm dns6conv
.Nd convert an IPv6 textual address for a DNS RR
.\"
.Sh SYNOPSIS
.Nm
.Op Fl b Ar beginbit
.Op Fl e Ar endbit
.Op Fl f Ar formattype
.Ar IPv6address
.\"
.Sh DESCRIPTION
.Nm
converts a specified textual IPv6 address into an appropriate DNS
resource record format.
It first separates a portion of the given address from the
.Ar beginbit
bit to the
.Ar endbit
bit, with
.Ar beginbit
inclusive and
.Ar endbit
exclusive
.Pq the 0th bit is the most significant .
The default values of theses bits are 64 and 128,
and can be specified by the
.Fl b
and
.Fl e
command line options, respectively.
When specified, the values must be between 0 and 128, with these two
values inclusive.
Also,
.Ar beginbit
must be smaller than
.Ar endbit .
.Pp
Then the command prints the separated address in an appropriate format
to be used as
.Pq a part of
a DNS resource record.
The
.Fl f
option is used to specify the type of the desired resource record.
The following values can be specified with the option.
.Bl -tag -width Ds -compact
.It Ic nibble
specifies to print the address as nibble DNS labels.
This is the default format type of this command.
When this format is specified,
.Ar beginbit
and
.Ar endbit
must be a multiple of 4.
.It Ic a6
specifies to print the separated address as an ordinary IPv6 textual
format, which can be used as the address suffix part of an A6 resource
record.
.It Ic bit
specifies to print the address as a hex string with the length of the
significant bits separated by the slash character.
The output can be used as a bitstring label to construct reverse map
database.
.El
.Sh RETURN VALUES
.Nm
exits with 0 on success, and non-zero on errors.
.\"
.Sh HISTORY
The
.Nm
command first appeared in the WIDE/KAME IPv6 protocol stack kit.


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/dns6conv/dns6conv.c
/*	$KAME: dns6conv.c,v 1.11 2002/04/10 09:04:36 jinmei Exp $ */

/*
 * Copyright (C) 2001 WIDE Project.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
#include <sys/types.h>
#include <sys/socket.h>

#include <netinet/in.h>

#include <stdio.h>
#include <unistd.h>
#include <netdb.h>
#include <err.h>
#include <stdlib.h>
#include <string.h>

enum {default_beg_bit = 64, default_end_bit = 128}; 
static enum {a6, bitlabel, nibble} fmttype;

static struct sockaddr_in6 *cut __P((const struct sockaddr_in6 *, int, int));
static char *cut_bitlabel __P((const struct in6_addr *, int, int));
static void print_bitstring __P((const char *, int));
static void print_nibble __P((const char *, int));
static char *fmt_t2s __P((int));
static void usage __P((void));

int
main(argc, argv)
	int argc;
	char *argv[];
{
	char *addr;
	const char *fmtstr = NULL, *ap;
	int ch, beg, end, error;
	struct addrinfo hints, *res;
	struct sockaddr_in6 *bin6;
	char hostbuf[NI_MAXHOST];

	beg = default_beg_bit;
	end = default_end_bit;

	memset(&hints, 0, sizeof(hints));
	hints.ai_family = AF_INET6;
	hints.ai_flags = AI_NUMERICHOST;

	while ((ch = getopt(argc, argv, "b:e:f:")) != -1) {
		switch (ch) {
		case 'b':
			beg = atoi(optarg);
			break;
		case 'e':
			end = atoi(optarg);
			break;
		case 'f':
			fmtstr = optarg;
			break;
		default:
			usage();
			/* NOTREACHED */
		}
	}
	argc -= optind;
	argv += optind;

	if (argc != 1)
		usage();
	addr = argv[0];

	if (fmtstr != NULL) {
		if (strncasecmp(fmtstr, "a6", 2) == 0)
			fmttype = a6;
		else if (strncasecmp(fmtstr, "bit", 3) == 0)
			fmttype = bitlabel;
		else if (strncasecmp(fmtstr, "nibble", 6) == 0)
			fmttype = nibble;
		else
			errx(1, "unknown format type: %s", fmtstr);
	}
	else
		fmttype = nibble;

	if ((error = getaddrinfo((const char *)addr, NULL, &hints, &res))
	    != 0)
		errx(1, "getaddrinfo(%s): %s", addr, gai_strerror(error));

	/* bit length validation: 0 <= beg < end <= 128 */
	if (beg < 0 || end < 0 || beg > 128 || end > 128 || beg > end)
		errx(1, "beginbit(%d) or endbit(%d) are invalid", beg, end);

	fprintf(stderr,
		"dns6conv: translate %s into the %s format, from %d to %d\n",
		addr, fmt_t2s(fmttype), beg, end);

	switch (fmttype) {
	case a6:
		if ((bin6 = cut((const struct sockaddr_in6 *)res->ai_addr,
				beg, end)) == NULL)
			exit(1);

		if ((error = getnameinfo((struct sockaddr *)bin6,
					 sizeof(*bin6), hostbuf,
					 sizeof(hostbuf), NULL, 0,
					 NI_NUMERICHOST)) != 0)
			errx(1, "getnameinfo failed: error = %d", error);

		printf("%s\n", hostbuf);
		break;
	case bitlabel:
		ap = cut_bitlabel(&((struct sockaddr_in6 *)res->ai_addr)->sin6_addr,
				  beg, end);
		print_bitstring(ap, end - beg);
		break;
	case nibble:
		if ((beg % 4) != 0 || (end % 4) != 0)
			errx(1, "beginbit(%d) or endbit(%d) is invalid "
			     "for the nibble format", beg, end);
		ap = cut_bitlabel(&((struct sockaddr_in6 *)res->ai_addr)->sin6_addr,
				  beg, end);
		print_nibble(ap, end - beg);
		break;
	}

	exit(0);
}

static struct sockaddr_in6 *
cut(src, beg, end)
	const struct sockaddr_in6 *src;
	int beg, end;
{
	static struct sockaddr_in6 ret;
	char *cp;
	const char *bp;
	int i, bbyte, ebyte;

	memset(&ret, 0, sizeof(ret)); /* just for safety */
	ret = *src;
	bp = (const char *)&src->sin6_addr;
	cp = (char *)&ret.sin6_addr;

	/*
	 *      <------- beg ------>
	 *      <--- bbyte --->
	 * src: xxxxxxxxxxxxxxxxxxxxBBaaaaaaaaaaaaaaEEyyyyyyyyyyy
	 * dst: 00000000000000000000BBaaaaaaaaaaaaaaEE00000000000
	 *      <---------------- end --------------->
	 *      <------------- ebyte -------------->
	 */
	bbyte = (beg & ~7) / 8;
	ebyte = (end & ~7) / 8;

	for (i = 0; i < bbyte; i++)
		cp[i] = 0;
	cp[bbyte] = (bp[bbyte] & (0x00ff >> (beg % 8))) & 0xff;
	if ((end % 8) != 0)
		cp[ebyte] = (bp[ebyte] & (0xff << (8 - (end % 8))));
	for (i = ebyte + 1; i < 16; i++)
		cp[i] = 0;

	return(&ret);
}

static char *
cut_bitlabel(a6, beg, end)
	const struct in6_addr *a6;
	int beg, end;
{
	static char ret[sizeof(struct in6_addr)];
	const u_char *bp = (const u_char *)a6;
	char *cp = ret;
	int bbyte, resid, sft;

	bbyte = (beg & ~7) / 8;
	bp += bbyte;
	resid = end - beg;
	sft = beg % 8;
	memset(ret, 0, sizeof(ret));

	while (resid > 0) {
		*cp = ((*bp << sft) & (0xff << sft)) & 0xff;
		if ((resid -= (8 - sft)) <= 0)
			break;
		bp++;
		*cp |= ((*bp >> (8 - sft)) & 0xff);
		cp++;
		resid -= sft;
	}

	return(ret);
}

static void
print_bitstring(cp, blen)
	const char *cp;
	int blen;
{
	char pbuf[NI_MAXHOST]; /* XXX */
	char *dn = pbuf, tc;
	char *ep = &pbuf[sizeof(pbuf)];
	int b;
	int n;

	for (b = blen; b > 7; b -= 8, cp++) {
		n = snprintf(dn, ep - dn, "%02x", *cp & 0xff);
		if (n < 0 || n >= ep - dn)
			goto fail;
		dn += n;
	}
	if (b > 4) {
		tc = *cp++;
		n = snprintf(dn, ep - dn, "%02x", tc & (0xff << (8 - b)));
		if (n < 0 || n >= ep - dn)
			goto fail;
		dn += n;
	} else if (b > 0) {
		tc = *cp++;
		n = snprintf(dn, ep - dn, "%1x",
		    ((tc >> 4) & 0x0f) & (0x0f << (4 - b))); 
		if (n < 0 || n >= ep - dn)
			goto fail;
		dn += n;
	}
	n = snprintf(dn, ep - dn, "/%d", blen);
	if (n < 0 || n >= ep - dn)
		goto fail;
	dn += n;

	printf("%s\n", pbuf);
	return;

fail:
	printf("?\n");
}

static void
print_nibble(cp0, blen)
	const char *cp0;
	int blen;
{
	int pbyte = blen / 8;
	const char *cp = cp0 + pbyte;

	if ((blen % 8) != 0) {
		printf("%1x%s", (*cp & 0xf0) / 16, blen > 4 ? "." : "");
		blen -= 4;
	}

	for (; blen > 0; blen -= 8) {
		cp--;
		printf("%1x.%1x%s", *cp & 0x0f, ((*cp) & 0xf0) / 16,
		       blen > 8 ? "." : "");
	}

	printf("\n");

	return;
}

static char *
fmt_t2s(int fmttype)
{
	switch (fmttype) {
	case a6:
		return("a6");
	case bitlabel:
		return("bitlabel");
	case nibble:
		return("nibble");
	}

	return(NULL);
}

static void
usage()
{
	fprintf(stderr,
		"usage: dns6conv [-b begbit] [-e endbit] "
		"[-f format] IPv6address\n");

	exit(1);
}
